src/ucp/tag/rndv.c

```diff
@@ -644,6 +659,77 @@ static void ucp_rndv_send_frag_rtr(ucp_worker_h worker, ucp_request_t *rndv_req,
     ucp_request_put(rndv_req);
 }
 
+static UCS_F_ALWAYS_INLINE int
+ucp_rndv_is_rkey_ptr(const ucp_rndv_rts_hdr_t *rndv_rts_hdr, ucp_ep_h ep,
+                     ucs_memory_type_t recv_mem_type, ucp_rndv_mode_t rndv_mode)
+{
+    const ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+    return /* must have remove address */
+           (rndv_rts_hdr->address != 0) &&
+           /* remove key must be on a memory domain for which we support rkey_ptr */
+           (ucp_rkey_packed_md_map(rndv_rts_hdr + 1) &
+            ep_config->tag.rndv.rkey_ptr_dst_mds) &&
+           /* rendezvous mode must not be forced to put/get */
+           (rndv_mode == UCP_RNDV_MODE_AUTO) &&
+           /* need local memory access for data unpack */
+           UCP_MEM_IS_ACCESSIBLE_FROM_CPU(recv_mem_type);
+}
+
+static void ucp_rndv_do_rkey_ptr(ucp_request_t *rndv_req, ucp_request_t *rreq,
+                                 const ucp_rndv_rts_hdr_t *rndv_rts_hdr)
+{
+    ucp_ep_h ep                      = rndv_req->send.ep;
+    const ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+    ucp_md_index_t dst_md_index;
+    ucp_lane_index_t i, lane;
+    ucs_status_t status;
+    unsigned rkey_index;
+    void *local_ptr;
+    ucp_rkey_h rkey;
+
+    ucp_trace_req(rndv_req, "start rkey_ptr rndv rreq %p", rreq);
+
+    status = ucp_ep_rkey_unpack(ep, rndv_rts_hdr + 1, &rkey);
+    if (status != UCS_OK) {
+        ucs_fatal("failed to unpack rendezvous remote key received from %s: %s",
+                  ucp_ep_peer_name(ep), ucs_status_string(status));
+    }
+
+    /* Find a lane which is capable of accessing the destination memory */
+    lane = UCP_NULL_LANE;
+    for (i = 0; i < ep_config->key.num_lanes; ++i) {
+        dst_md_index = ep_config->key.lanes[i].dst_md_index;
+        if (UCS_BIT(dst_md_index) & rkey->md_map) {
+            lane = i;
+            break;
+        }
+    }
+
+    if (ucs_unlikely(lane == UCP_NULL_LANE)) {
+        /* We should be able to find a lane, because ucp_rndv_is_rkey_ptr()
+         * already checked that (rkey->md_map & ep_config->rkey_ptr_dst_mds) != 0
+         */
+        ucs_fatal("failed to find a lane to access remote memory domains 0x%lx",
+                  rkey->md_map);
+    }
+
+    rkey_index = ucs_bitmap2idx(rkey->md_map, dst_md_index);
+    status     = uct_rkey_ptr(rkey->tl_rkey[rkey_index].cmpt,
+                              &rkey->tl_rkey[rkey_index].rkey,
+                              rndv_rts_hdr->address, &local_ptr);
+    if (status == UCS_OK) {
+        ucp_trace_req(rndv_req, "obtained a local pointer to remote buffer: %p",
+                      local_ptr);
+        status = ucp_request_recv_data_unpack(rreq, local_ptr,
+                                              rndv_rts_hdr->size, 0, 1);
+        ucp_request_complete_tag_recv(rreq, status);
+    }
+
+    ucp_rkey_destroy(rkey);
+    ucp_rndv_req_send_ats(rndv_req, rreq, rndv_rts_hdr->sreq.reqptr);
```

Comment:
should not we send actual status if something goes wrong?
Currently `ucp_rndv_req_send_ats` always sends `UCS_OK`

Reply:
fixed